_posts/2023-09-07-develop-functions-locally.md

```diff
@@ -0,0 +1,278 @@
+
+## Use secrets with local-run
+
+All functions can consume secrets in the same way, by reading a file from: `/var/openfaas/secrets/NAME`
+
+To mount a secret in a function the secret name has to be added to the list of secrets in the stack YAML file.
+
+As an example we will add a secret named `api-key` to the echo function:
+
+```yaml
+functions:
+  echo:
+    lang: node18
+    handler: ./echo
+    image: echo:latest 
+    secrets:
+      - api-key
+```
+
+The local-run command looks for secret files in the `.secrets` folder. You will need to create any secrets you want in this location.
+
+All secrets included in a functions stack.yaml will be mounted into the function container so they can be read from their usual location,`/var/openfaas/secrets/NAME, and used within the function.
+
+Create the `.secrets` folder in your local directory and add a file named api-key.
+
+```bash
+mkdir .secrets
+echo "secret-acces-token" > .secrets/api-key
+```
+
+Update the echo function to read the secret file from `/var/openfaas/secrets/api-key` and add the secret value to the response:
+
+```js
+'use strict'
+
+const { readFile } = require('fs').promises
+
+module.exports = async (event, context) => {
+  let token = await readFile("/var/openfaas/secrets/api-key", )
+
+  const result = {
+    'body': JSON.stringify(event.body),
+    'api-key': token.toString(),
+    'content-type': event.headers["content-type"]
+  }
+
+  return context
+    .status(200)
+    .succeed(result)
+}
+```
+
+Run the function with local-run and see that the secret is mounted in the container and can be read for use within the function.
+
+Run the echo function:
+
+```bash
+faas-cli local-run echo
+```
+
+Invoke the function with curl and inspect the response:
+
+```bash
+$ curl -s http://127.0.0.1:8080 \
+  -H "Content-Type: text/plain" \
+  -d "Greetings from OpenFaaS" | jq
+
+{
+  "body": "\"Greetings from OpenFaaS\"",
+  "api-key": "secret-access-token",
+  "content-type": "text/plain"
+}
+```
+
+You should see that the api-key secret is included in the JSON response.
+
+## Watch for changes
+
+Another way to quickly iterate on function during development is by using the built in watch functionality of the CLI.
+
+The `--watch` flag can be used with both the `local-run` and `up` command. Adding the flag will tell the cli to watch for any changes to the function source files and automatically re-build and deploy functions on save.
+
+## Automatic tagging of function images
+
+All OpenFaaS functions are built into immutable docker images. By default if no image tag is included for a function in the stack.yml file the `:latest` tag is used. When iterating over function and pushing them to an image registry it is a best practice to organise different image versions using tags instead of always pushing to `:latest`.
+
+There are two options to set tags for function images.
+
+1. Set the image tag in the stack.yml file
+    
+    ```
```

Comment:
This indentation is incorrect.

Copy the example from earlier to show where this line is placed in the file.